internal/deviceplugin/server.go

```diff
@@ -24,38 +24,144 @@ import (
+	updatesCh    chan map[string]DeviceInfo
+	devices      map[string]DeviceInfo
+	postAllocate func(*pluginapi.AllocateResponse) error
 }
 
-// Serve serves starts gRPC server to serve Device Plugin functionality
-func (srv *Server) Serve(dm pluginapi.DevicePluginServer, resourceName string, pluginPrefix string) error {
-	return srv.setupAndServe(dm, resourceName, pluginPrefix, pluginapi.DevicePluginPath, pluginapi.KubeletSocket)
+// newServer creates a new server satisfying the devicePluginServer interface.
+func newServer(devType string, postAllocate func(*pluginapi.AllocateResponse) error) devicePluginServer {
+	return &server{
+		devType:      devType,
+		updatesCh:    make(chan map[string]DeviceInfo, 1), // TODO: is 1 needed?
+		devices:      make(map[string]DeviceInfo),
+		postAllocate: postAllocate,
+	}
+}
+
+func (srv *server) GetDevicePluginOptions(ctx context.Context, empty *pluginapi.Empty) (*pluginapi.DevicePluginOptions, error) {
+	fmt.Println("GetDevicePluginOptions: return empty options")
+	return new(pluginapi.DevicePluginOptions), nil
+}
+
+func (srv *server) sendDevices(stream pluginapi.DevicePlugin_ListAndWatchServer) error {
+	resp := new(pluginapi.ListAndWatchResponse)
+	for id, device := range srv.devices {
+		resp.Devices = append(resp.Devices, &pluginapi.Device{id, device.State})
+	}
+	glog.V(2).Info("Sending to kubelet ", resp.Devices)
+	if err := stream.Send(resp); err != nil {
+		srv.Stop()
+		return fmt.Errorf("device-plugin: cannot update device list: %v", err)
+	}
+
+	return nil
 }
 
-// Stop stops serving Device Plugin
-func (srv *Server) Stop() error {
+func (srv *server) ListAndWatch(empty *pluginapi.Empty, stream pluginapi.DevicePlugin_ListAndWatchServer) error {
+	glog.V(2).Info("Started ListAndWatch for ", srv.devType)
+
+	if err := srv.sendDevices(stream); err != nil {
+		return err
+	}
+
+	for srv.devices = range srv.updatesCh {
+		if err := srv.sendDevices(stream); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (srv *server) Allocate(ctx context.Context, rqt *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	response := new(pluginapi.AllocateResponse)
+	for _, crqt := range rqt.ContainerRequests {
+		cresp := new(pluginapi.ContainerAllocateResponse)
+		for _, id := range crqt.DevicesIDs {
+			dev, ok := srv.devices[id]
+			if !ok {
+				return nil, fmt.Errorf("Invalid allocation request with non-existing device %s", id)
+			}
+			if dev.State != pluginapi.Healthy {
+				return nil, fmt.Errorf("Invalid allocation request with unhealthy device %s", id)
+			}
+			for _, devnode := range dev.Nodes {
+				cresp.Devices = append(cresp.Devices, &pluginapi.DeviceSpec{
+					HostPath:      devnode,
+					ContainerPath: devnode,
+					Permissions:   "mrw",
+				})
+			}
+			for _, devmount := range dev.Mounts {
+				cresp.Mounts = append(cresp.Mounts, &pluginapi.Mount{
+					HostPath:      devmount,
+					ContainerPath: devmount,
+					ReadOnly:      false,
+				})
+			}
+		}
```

Comment:
A note to myself: I forgot setting the cresp.Envs map on this cycle, so QAT is broken now.